### PR TITLE
Web - V1

### PR DESCRIPTION
**Pull request by mikejahn**
* Web exercise to read in names and display list of names. Also stored them in a file so names are not lost on shutdown. Read in files on startup

### CODE REVIEW

File: exercise-003-web/exercise/server.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		log.Fatal(err)
+	}
+	data, err := ioutil.ReadAll(file)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	userNames = strings.Split(string(data), "\n")
+	// Remove extra line
+	if userNames[len(userNames)-1] == "" {
```
Comment (wmbest2) [inline exercise-003-web/exercise/server.go]: You could use `strings.TrimSpace` for this before the split

File: exercise-003-web/exercise/server.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+const fileName = "names.txt"
+
+var indexT = template.Must(template.ParseFiles("./index.html"))
+var userNames []string
+
+func index(w http.ResponseWriter, r *http.Request) {
+	v := View{UserNames: userNames}
+	indexT.Execute(w, &v)
```
Comment (spectreman) [inline exercise-003-web/exercise/server.go]: This will list all the names. I believe the exercise asks for a list of names with their counts.

**Comment by spectreman** [comment]
The output on the web-page should be something like:

```
Timmy 2
Tommy 1
...
```

File: exercise-003-web/exercise/server.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+// View is the data object passed into template
+type View struct {
+	UserNames []string
+}
+
+const fileName = "names.txt"
+
+var indexT = template.Must(template.ParseFiles("./index.html"))
```
Comment (wmbest2) [inline exercise-003-web/exercise/server.go]: These can also be formatted as 

~~~go
var (
    indexT = template.Must(template.ParseFiles("./index.html"))
    userNames []string
)
~~~

File: exercise-003-web/exercise/server.go (hunk 2 above)
Comment (mikejahn) [reply exercise-003-web/exercise/server.go]: @spectreman ah yes, you are correct. I mis-read the description. I'll open a new PR with that updated!

File: exercise-003-web/exercise/server.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+func signup(w http.ResponseWriter, r *http.Request) {
+	r.ParseForm()
+	username := r.Form.Get("username")
+	userNames = append(userNames, username)
+	writeToFile()
+	http.Redirect(w, r, "/", 301)
+}
+
+func writeToFile() {
```
Comment (wmbest2) [inline exercise-003-web/exercise/server.go]: Totally a personal preference for something this small but I would prefer write and read to have the function headers 

~~~go
func writeUsernamesToFile(userNames []string, filename string) (error)
~~~

and 

~~~go
func readUsernameFile(filename string) ([]string, error)
~~~

**Pull request closed without merging**